USHIFT 1104 - MicroShift CSI Snapshotting Integration (#1406)
openshift/enhancements: enhancements/microshift/feature-csi-snapshots.md

## Motivation

CSI snapshot functionality was originally excluded from the CSI driver integration in MicroShift to support the low-resource overhead goals of the project. However, user feedback has made it clear that a supportable, robust backup/restore solution is necessary.  While it would be possible to run a workflow out-of-band to manage workload data, this would be reinventing the wheel and contribute significantly to technical debt.  CSI Snapshots are already integrated into Openshift with ongoing downstream support. 

### User Stories

As an edge device owner running stateful workloads on MicroShift, I want to create in-cluster snapshots of that state and to restore workloads to that state utilizing existing Kubernetes patterns.

### Goals

* Enable an in-cluster workflow for snapshotting and restoration of cluster workload data
* Follow the [MicroShift design principles](https://github.com/openshift/microshift/blob/main/docs/design.md)

### Non-Goals

* Enable backup/restore workflows outside of MicroShift

> * Provide a workflow for exporting data outside of a MicroShift cluster

Reviewer: I think we also said we wouldn't worry about making this optional, for now, that we would address that holistically in another design.

## Proposal

Deploy the CSI snapshot controller and CSI plugin sidecar with MicroShift out of the box to provide users with a means of snapshotting, cloning, and restoring workload state.  Resource conscious users should be able to opt out of deploying these components.

### Workflow Description

#### Deploy MicroShift with Snapshotting

_Assuming_
* A running MicroShift cluster

> * A running workload with an attached volume, backed by an LVM thin volume

Reviewer: This assumed starting point doesn't make as much sense for this deployment workflow as it does for some of the other workflows. Did you mean for it to be here, or was it copy-and-pasted?

## Motivation

CSI snapshot functionality was originally excluded from the CSI driver integration in MicroShift to support the low-resource overhead goals of the project. However, user feedback has made it clear that a supportable, robust backup/restore solution is necessary.  While it would be possible to run a workflow out-of-band to manage workload data, this would be reinventing the wheel and contribute significantly to technical debt.  CSI Snapshots are already integrated into Openshift with ongoing downstream support. 

### User Stories

As an edge device owner running stateful workloads on MicroShift, I want to create in-cluster snapshots of that state and to restore workloads to that state utilizing existing Kubernetes patterns.

### Goals

* Enable an in-cluster workflow for snapshotting and restoration of cluster workload data
* Follow the [MicroShift design principles](https://github.com/openshift/microshift/blob/main/docs/design.md)

### Non-Goals

* Enable backup/restore workflows outside of MicroShift
* Provide a workflow for exporting data outside of a MicroShift cluster

> ## Proposal

Author: I've updated the design details section to include a complete list of additional components and a list of the manifest files that will be added to MicroShift to define those components.

The VolumeSnapshotClass CRD defines the parameters for the csi-external-snapshotter sidecar to use when creating a snapshot. This allows the storage back end to know what kind of snapshot to dynamically create if multiple options are supported.

Dynamically provisioned snapshots use the VolumeSnapshotClass CRD to specify storage-provider-specific parameters to use when creating a snapshot.

The VolumeSnapshotContentClass CRD is not namespaced and is for use by a cluster administrator to enable global configuration options for their storage back end.

```yaml
apiVersion: snapshot.storage.k8s.io/v1
kind: VolumeSnapshotClass
metadata:
  name: csi-hostpath-snap
driver: hostpath.csi.k8s.io [1] 
deletionPolicy: Delete
```

1. The name of the CSI driver that is used to create snapshots of this **VolumeSnapshotClass** object. The name must be the same as the Provisioner field of the storage class that is responsible for the PVC that is being snapshotted.

#### VolumeSnapshot

> Similar to the PersistentVolumeClaim object, the VolumeSnapshot CRD defines a developer request for a snapshot. The CSI Snapshot Controller Operator runs the CSI snapshot controller, which handles the binding of a VolumeSnapshot CRD with an appropriate VolumeSnapshotContent CRD. The binding is a one-to-one mapping.

Reviewer: Are we bringing the CSI Snapshot Controller Operator into MicroShift?

Author: C/p error from OpenShift documentation, will drop.

spec:
  deletionPolicy: Delete
  driver: hostpath.csi.k8s.io
  source:
    snapshotHandle: 7bdd0de3-aaeb-11e8-9aae-0242ac110002
  sourceVolumeMode: Filesystem
  volumeSnapshotRef:
    name: new-snapshot-test
    namespace: default
```

### Risks and Mitigations

- Increased overhead.  The CSI snapshot controller and sidecar increase the on-disk footprint by a
total of 360Mb.  At idle, the components use a negligible amount of CPU (>1% CPU on a 2 core machine)
and roughly 18Mb.  Under load, the 

### Drawbacks

## Design Details

### Test Plan

Snapshotting functionality will be tested under the Openshift conformance test suite.

### Graduation Criteria

#### Dev Preview -> Tech Preview

#### Tech Preview -> GA

#### Full GA

- Available by default

#### Removing a deprecated feature

### Upgrade / Downgrade Strategy

The CSI Snapshot controller and sidecar are pieces of the Kubernetes CSI implementation.  OpenShift
maintains downstream versions of these images and tracks them as part of OCP releases.

### Version Skew Strategy

The CSI Snapshot version is tracked by the ocp-release image.  This provides sufficient guardrails

> against version skew.

Author: The CRDs are versioned with Kube.

> against version skew.

Reviewer: Does that mean we have to worry about the kube rebase being applied to cluster-csi-snapshot-controller-operator before MicroShift can take the same update?

> against version skew.

Reviewer: I think I'm not communicating my concern clearly.

I can't tell which repository the CRDs for these new types are going to come from. I'm worried that if they are not in the `openshift/kubernetes` repository then when OpenShift updates its kube version we will have some version skew between components that we are vendoring. We already have a similar situation with the route-controller-manager that means MicroShift cannot update its kube version until the RCM does. Is this a new case that will require updating something else, like a storage-related operator?

> against version skew.

Author: Is this only in the vendoring context? The design here would be to deploy the csi-controller as a pod on the cluster, not to embed it.  We wouldn't be rebasing code, just updating the image digests.

I see the CRDs are also defined in [openshift/kubernetes](https://github.com/openshift/kubernetes/tree/15f19ea2dd700767e5337502aec753d2a6e26905/cluster/addons/volumesnapshots/crds), so we could pull them from here instead to stay on the safe side.

> against version skew.

Reviewer: Regardless of how we run the component, we need to bring the manifests into the MicroShift repository and we want them to match the code that is being run.

> against version skew.

Author: Yeah, understood.  We should stick with sourcing them from the openshift-csi-controller-operator in that case.  Reason being that the openshift/kubernetes CRDs are versioned with the upstream container image [here](https://github.com/openshift/kubernetes/blob/15f19ea2dd700767e5337502aec753d2a6e26905/cluster/addons/volumesnapshots/volume-snapshot-controller/volume-snapshot-controller-deployment.yaml#LL25C9-L25C9), which we don't want to use since Openshift already maintains a downstream.  The controller-operator is versioned with the Openshift versions of the images from quay.io/openshift4/, [here](https://github.com/openshift/cluster-csi-snapshot-controller-operator/blob/eaaf7260464bd6cb6126dfd7259cb6217eb25340/manifests/07_deployment.yaml#L42-L49).

> against version skew.

Author: Yes, that's what I was thinking.  Wherever we get the CRD manifests from, we should also use the relative image, as they are versioned together.  That way, we can be pretty certain that the CRDs and the container versions are synced.

CSI snapshot functionality was originally excluded from the CSI driver integration in MicroShift to
support the low-resource overhead goals of the project. However, user feedback has made it clear that
a supportable, robust backup/restore solution is necessary.  While it would be possible to run a 
workflow out-of-band to manage workload data, this would be reinventing the wheel and contribute
significantly to technical debt.  CSI Snapshots are already integrated into Openshift with ongoing
downstream support. 

### User Stories

As an edge device owner running stateful workloads on MicroShift, I want to create in-cluster snapshots of that state 
and to restore workloads to that state utilizing existing Kubernetes patterns.

### Goals

* Enable an in-cluster workflow for snapshotting and restoration of cluster workload data
* Follow the [MicroShift design principles](https://github.com/openshift/microshift/blob/main/docs/design.md)

### Non-Goals

* Enable backup/restore workflows external to MicroShift
* Provide a workflow for exporting data from a MicroShift cluster
* Provide a means of opting out.  This falls under platform composability and is out of scope

## Proposal

> Deploy the CSI snapshot controller and CSI plugin sidecar with MicroShift out of the box to provide users with a means of snapshotting, cloning, and restoring workload state.  Resource conscious users should be able to opt out of deploying these components.

Author: I think instead I'll drop this sentence entirely, since composability is out of scope of the enhancement.

> Deploy the CSI snapshot controller and CSI plugin sidecar with MicroShift out of the box to provide users with a means of snapshotting, cloning, and restoring workload state.  Resource conscious users should be able to opt out of deploying these components.

Reviewer: Dropping it also makes sense.

### Workflow Description

#### Deploy MicroShift with Snapshotting

_Workflow_

1. Install MicroShift RPMs
2. Start the MicroShift systemd service
3. Observe the CSI Snapshot controller pod reaches the Ready state
4. Observe the topolvm-controller and csi-snapshot-controller pods reach the Ready state

> #### Snapshot, Dynamic

Reviewer: If you added a quick sentence to each workflow expanding on the name, that would make this section clearer.

_Assuming_

* A running MicroShift cluster
* A running workload with an attached volume, backed by an LVM thin volume

_Workflow_

1. A user creates a VolumeSnapshot, specifying the source PVC and VolumeSnapshotClass
2. The CSI Snapshot Controller creates a VolumeSnapshotContent instance
3. The snapshot driver creates a new backing volume and clones the data from the source PVC’s volume
4. The CSI Snapshot Controller binds the VolumeSnapshotContent and the VolumeSnapshot together, signaling completion
5. The VolumeSnapshot may now be referenced by PVCs as a data source

#### Snapshot, Static

_Assuming_

* A running MicroShift cluster
* A pre-provisioned VolumeSnapshotContent API obj, representing a backend volume containing pre-populated data

> * The CSI Snapshot controller is not deployed

Author: If the controller were not deployed, then yes, this workflow makes sense. That said, I'm going to adjust this workflow to something that makes a bit more sense in the context.  It shouldn't assume a pre-provisioned volume, for instance.

> * The CSI Snapshot controller is not deployed

Reviewer: There's some confusion about whether this feature will be optional. In other threads, we're converging on "not optional, for now". Maybe this is just another spot that needs to be updated with that in mind?

_Workflow_

1. The user creates a VolumeSnapshot, specifying the VolumeSnapshotContent name as the source
2. The user creates a PVC, specifying the VolumeSnapshot as the dataSource.
3. The storage driver creates a new backing volume and clones the data from the dataSource to the new volume
4. The VolumeSnapshot will is now available as a PVC data source.

#### Restore

_Assuming_

* A running MicroShift cluster
* A bound VolumeSnapshot

_Workflow_

1.  A user creates a Pod and PVC, specifying the VolumeSnapshot as the data source
2. The CSI storage driver (topolvm), creates a new backing thin volume and a PV to represent the volume
3. The CSI storage driver creates a PV to expose the backend storage at the cluster level
4. The CSI storage driver clones the snapshot volume data to the new volume
5. The in-tree storage controller binds the PV to the PVC
6. The volume is mounted to the Pod’s filesystem and the Pod is started

#### Deploying

The CSI Snapshot Controller and TopoLVM components are deployed by default on MicroShift. The

> manifests for these components are baked into the MicroShift binary and are deployed upon first-boot.

Author: Ah, true. Will fix.

```yaml
apiVersion: snapshot.storage.k8s.io/v1
kind: VolumeSnapshotContent
metadata:
  name: new-snapshot-content-test
  annotations:
    - snapshot.storage.kubernetes.io/allow-volume-mode-change: "true"
spec:
  deletionPolicy: Delete
  driver: hostpath.csi.k8s.io
  source:
    snapshotHandle: 7bdd0de3-aaeb-11e8-9aae-0242ac110002
  sourceVolumeMode: Filesystem
  volumeSnapshotRef:
    name: new-snapshot-test
    namespace: default
```

### Risks and Mitigations

- Increased overhead.  The CSI snapshot controller and sidecar increase the on-disk footprint by a
total of 360Mb.  At idle, the components use a negligible amount of CPU (>1% CPU on a 2 core machine)
and roughly 18Mb.

### Drawbacks

## Design Details

### CSI Snapshotter Components

As mentioned above, CSI Volume Snapshot logic is divided between two runtimes: the CSI Volume Snapshot Controller 
and the CSI Volume Snapshot sidecar.  These runtimes are pre-packaged in container images and available through the
OpenShift container registry.  Additionally, a validating webhook verifies the correctness of CRD instances.

The total additional cluster components are:

> 1. CSI VolumeSnapshot Controller Deployment

Reviewer: This will produce a new pod?

Author: Yes, 1 pod running 1 container internally.

> 2. CSI VolumeSnapshot Validating Webhook Deployment, responsible for validating CRD instances

Reviewer: Where will the web hook run? Is it part of the pod created by the deployment mentioned in item 1?

#### Removing a deprecated feature

### Upgrade / Downgrade Strategy

**Images**

The CSI Snapshot controller and sidecar are pieces of the Kubernetes CSI implementation.  OpenShift
maintains downstream versions of these images and tracks them as part of OCP releases.

**Manifests**

MicroShift's rebase automation specifies repos in the OpenShift github organization from which specific manifests are obtained.
For the CSI Snapshot Controller, we will specify https://github.com/openshift/cluster-csi-snapshot-controller-operator/tree/release-$RELEASE/assets
as the remote source from which to derive the controller manifests.

LVMS does not provide its manifests as plain yaml files.  Instead, these are encoded into the logic of the operator, which
is not deployed on MicroShift.  This makes retrieving them automatically difficult.  For this reason, the LVMS manifests 
are derived once from a running instance of the controller and stored under `microshift/assets/components/lvms/`.

### Version Skew Strategy

The CSI Snapshot version is tracked by the ocp-release image.  This provides sufficient guardrails
against version skew.

> ### Operational Aspects of API Extensions

Reviewer: We are adding a few new APIs so we should fill in these sections about how they might fail and how someone is supposed to debug the failures when something goes wrong. Are there specific CRs or log files to examine for errors, for example?

Can the new web hooks introduce new failure modes?

Author: 👍🏻

- More testing (upgrade, downgrade, scale)
- Sufficient time for feedback
- Available by default
- Document SLOs for the component
- Conduct load testing
- User facing documentation created in [openshift-docs](https://github.com/openshift/openshift-docs/)

#### Removing a deprecated feature

### Upgrade / Downgrade Strategy

**Images**

The CSI Snapshot controller and sidecar are pieces of the Kubernetes CSI implementation.  OpenShift
maintains downstream versions of these images and tracks them as part of OCP releases.

**Manifests**

MicroShift's rebase automation specifies repos in the OpenShift github organization from which specific manifests 
are obtained. For the CSI Snapshot Controller, we will specify 
https://github.com/openshift/cluster-csi-snapshot-controller-operator/tree/release-$RELEASE/assets 
as the remote source from which to derive the controller manifests and image references.

### Version Skew Strategy

The CSI Snapshot version is tracked by the ocp-release image.  This provides sufficient guardrails
against version skew.

### Operational Aspects of API Extensions

#### Failure Modes

#### Support Procedures

## Implementation History

## Alternatives

> There are no alternatives to deploying the Kubernetes CSI implementation of the snapshotting specification.

Reviewer: We could not provide the feature.

and to restore workloads to that state utilizing existing Kubernetes patterns.

### Goals

* Enable an in-cluster workflow for snapshotting and restoration of cluster workload data
* Follow the [MicroShift design principles](https://github.com/openshift/microshift/blob/main/docs/design.md)

### Non-Goals

* Enable backup/restore workflows external to MicroShift
* Provide a workflow for exporting data from a MicroShift cluster
* Provide a means of opting out.  This falls under platform composability and is out of scope

## Proposal

Deploy the CSI snapshot controller and CSI plugin sidecar with MicroShift out of the box to provide users with a means of 
snapshotting, cloning, and restoring workload state.

### Workflow Description

#### Deploy MicroShift with Snapshotting Support

_Prerequisites_

* The device owner has created an LVM volume group and a thin-pool on this volume group.
* The `/etc/microshift/lvmd.yaml` config includes a `deviceClass` to represent the thin-pool.  The lvmd.yaml may 
contain a mix of thin and thick `deviceClasses`.

```yaml
device-classes:
- name: ssd-thin
  volume-group: myvg1
  type: thin
  thin-pool:
    name: pool0
    overprovision-ratio: 50.0

> ```

Author: ah, right. will add

_Workflow_

1. The user installs the MicroShift RPMs, either via os-rpmtree layer or package manager.
2. The user starts the MicroShift systemd service
3. MicroShift starts the Service Manager loop to begin deploying cluster resources.
4. The Service Manager initiates the `startCSIPlugin()` service.

> 5. The `startCSIPlugin()` service deploys the CSI VolumeSnapshot CSI Snapshot Controller and Validating Webhook, and a 

Author: Will include that in the doc.

points to a real snapshot in the storage back end.

For manually pre-provisioned snapshots, a cluster administrator creates a number of
VolumeSnapshotContent CRDs. These carry the details of the real volume snapshot in the storage system.

The VolumeSnapshotContent CRD is not namespaced and is for use by a cluster administrator.

```yaml
apiVersion: snapshot.storage.k8s.io/v1
kind: VolumeSnapshotContent
metadata:
  name: new-snapshot-content-test
  annotations:
    - snapshot.storage.kubernetes.io/allow-volume-mode-change: "true"
spec:
  deletionPolicy: Delete
  driver: hostpath.csi.k8s.io
  source:
    snapshotHandle: 7bdd0de3-aaeb-11e8-9aae-0242ac110002
  sourceVolumeMode: Filesystem
  volumeSnapshotRef:
    name: new-snapshot-test
    namespace: default
```

### Risks and Mitigations

- Increased overhead.  The CSI snapshot controller and sidecar increase the on-disk footprint by a
total of 360Mb.  At idle, the components use a negligible amount of CPU (>1% CPU on a 2 core machine)
and roughly 18Mb.

### Drawbacks

## Design Details

### CSI Snapshotter Components

The total additional cluster components are:

9. VolumeSnapshot CRD
10. VolumeSnapshotContent CRD

> 11. VolumeSnapshotClass CRD

Author: thx, will fix

the image references from the same source as the CRD manifests to ensure compatibility.

### Operational Aspects of API Extensions

#### Failure Modes

- _Not Enough Storage Capacity:_ If the volume group capacity has been reached, snapshotting can fail because the 
  storage driver refuses to create a snapshot.  This will be observable as the VolumeSnapshot being stuck in the 
  **Pending** phase.

- _Valification Failure_: Webhook validation checks for malformed VolumeSnapshots and will post errors to the 
  problem object as Events. 

_Troubleshooting_

To determine the cause of snapshot failures, use `oc describe` to examine VolumeSnapshot and VolumeSnapshotContent 
events. 

- **VolumeSnapshot:**  `$ oc describe volumesnapshot -n $NAMESPACE $NAME`
- **VolumeSnapshotContent:** `$ oc describe volumesnapshotconent $NAME`

Errors related to restoring a snapshot to a PVC will be captured in the PVC's events.

- **PersistentVolumeClaim:** `$ oc describe pvc -n $NAMESPACE $NAME`

If you need to delve deeper, the logs can be examined with the following command:

`$ oc logs -n kube-system csi-snapshot-controller-$SUFFIX`

> #### Support Procedures

Author: Will add those instructions as well

tracking-link:
  - https://issues.redhat.com/browse/USHIFT-1140
see-also: []
---

# CSI Snapshotting Integration

## Summary

MicroShift is a small form-factor, single-node OpenShift targeting IoT and Edge Computing use cases characterized 
by tight resource constraints, unpredictable network connectivity, and single-tenant workloads. See 
[kubernetes-for-devices-edge.md](./kubernetes-for-device-edge.md) for more detail.

This document proposes the integration of the CSI Snapshot Controller to support backup and restore scenarios for 
cluster workloads.  The snapshot controller, along with the CSI external snapshot sidecar, will provide an API 
driven pattern for managing stateful workload data.

## Motivation

CSI snapshot functionality was originally excluded from the CSI driver integration in MicroShift to support the 
low-resource overhead goals of the project. However, user feedback has made it clear that a supportable, robust 
backup/restore solution is necessary.  While it would be possible to run a workflow out-of-band to manage workload 
data, this would be reinventing the wheel and contribute significantly to technical debt.  CSI Snapshots are already 

> integrated into Openshift with ongoing downstream support. 

Reviewer 2: ```suggestion
integrated into OpenShift with ongoing downstream support. 
```

### User Stories

As an edge device owner running stateful workloads on MicroShift, I want to create in-cluster snapshots of that state 
and to restore workloads to that state utilizing existing Kubernetes patterns.

### Goals

* Enable an in-cluster workflow for snapshotting and restoration of cluster workload data
* Follow the [MicroShift design principles](https://github.com/openshift/microshift/blob/main/docs/design.md)

> ### Non-Goals

Reviewer 2: I would add another non-goal, which is to describe how MicroShift and its configuration is installed / configured on a device. This might help us focus the ER on the main topic. 
I left few more comments below in the places where installation-related steps can be removed if we agree that's a way to go.

* Enable backup/restore workflows external to MicroShift
* Provide a workflow for exporting data from a MicroShift cluster

> * Provide a means of opting out.  This falls under platform composability and is out of scope

Reviewer 2: The "optiing out" part is not immediately clear to me. Can we rephrase / elaborate here?

## Proposal

Deploy the CSI snapshot controller and CSI plugin sidecar with MicroShift out of the box to provide users with a means of 
snapshotting, cloning, and restoring workload state.

### Workflow Description

#### Deploy MicroShift with Snapshotting Support

_Prerequisites_

* The device owner has created an LVM volume group and a thin-pool on this volume group.
* The `/etc/microshift/lvmd.yaml` config includes a `deviceClass` to represent the thin-pool.  The lvmd.yaml may 
contain a mix of thin and thick `deviceClasses`.

```yaml
device-classes:
- name: ssd-thin
  volume-group: myvg1
  type: thin
  thin-pool:
    name: pool0
    overprovision-ratio: 50.0
- name: default
  volume-group: rhel
  default: true
  spare-gb: 0
```

_Workflow_

> 1. The user installs the MicroShift RPMs, either via os-rpmtree layer or package manager.

Reviewer 2: ```suggestion
1. The user installs the MicroShift RPMs, either via ostree layer or RPM  package manager.
```

2. The user starts the MicroShift systemd service

> 3. MicroShift starts the Service Manager loop to begin deploying cluster resources.

Reviewer 2: I think steps 1-to-3 can be collapsed into a single one as installation related aspects are not relevant for this proposal. Let's just describe what happens when MicroShift service starts?

components are baked into the MicroShift binary and are deployed upon startup. This follows the existing pattern 
for MicroShift’s control-plane elements.

The CSI Snapshotter configuration is managed via the VolumeSnapshotClass API. This API serves a similar purpose as 
StorageClasses and allows admins to specify dynamic snapshotting parameters at runtime.  MicroShift will deploy a 
default VolumeSnapshotClass on startup. This instance will reference the default StorageClass that is already 
deployed by MicroShift to enable snapshotting out of the box.

The MicroShift deployment model utilizes rpm-ostree layers. Following the existing deployment pattern, 
VolumeSnapshotClass manifests can be packaged and deployed onto target devices.
(See [kubernetes-for-device-edge. md#workflow-description.md](./kubernetes-for-device-edge.md#workflow-description)).
Application deployers can predefine VolumeSnapshotClass manifests and use image-builder to package them into a 
rpm-ostree layer.  This layer can be installed to the device, which writes the manifests to 
/etc/microshift/manifests. This pattern is congruent with how a device owner would install custom StorageClasses. It 

> is recommended that new custom StorageClasses be packaged with VolumeSnapshotClasses that reference them.

Reviewer 2: Here again, I wonder if we need to go into the details of how the manifests are deployed? The bottom line is that we suggest the users to put their manifests in `/etc/microshift/manifests` directory to be consumed by MicroShift on restart.

#### Upgrading

The CSI Snapshot component images are packaged and versioned with each OpenShift release and can be extracted from 
the ocp-release image. This allows MicroShift to use existing rebase tooling to upgrade the CSI Snapshot components 

> in step with OCP releases.

Reviewer 2: I think we might want to explain this little more because "use existing rebase tooling" assumes knowledge of how the MicroShift rebase works. Without going into much details, can we state explicitly why this works?

Reviewer: Maybe saying that we will update rebase.sh to include the new manifests is enough?

create a snapshot of. Required for dynamically provisioning a snapshot.

#### VolumeSnapshotContent

A snapshot taken of a volume in the cluster that has been provisioned by a cluster administrator.

Similar to the PersistentVolume object, the VolumeSnapshotContent CRD is a cluster resource that
points to a real snapshot in the storage back end.

For manually pre-provisioned snapshots, a cluster administrator creates a number of
VolumeSnapshotContent CRDs. These carry the details of the real volume snapshot in the storage system.

The VolumeSnapshotContent CRD is not namespaced and is for use by a cluster administrator.

```yaml
apiVersion: snapshot.storage.k8s.io/v1
kind: VolumeSnapshotContent
metadata:
  name: new-snapshot-content-test
  annotations:
    - snapshot.storage.kubernetes.io/allow-volume-mode-change: "true"
spec:
  deletionPolicy: Delete
  driver: hostpath.csi.k8s.io
  source:
    snapshotHandle: 7bdd0de3-aaeb-11e8-9aae-0242ac110002
  sourceVolumeMode: Filesystem
  volumeSnapshotRef:
    name: new-snapshot-test
    namespace: default
```

### Risks and Mitigations

- Increased overhead.  The CSI snapshot controller and sidecar increase the on-disk footprint by a

> total of 360Mb.  At idle, the components use a negligible amount of CPU (>1% CPU on a 2 core machine)

Reviewer 2: ```suggestion
total of 360MB.  At idle, the components use a negligible amount of CPU (>1% CPU on a 2 core machine)
```

> and roughly 18Mb.

Reviewer 2: How much is 360MB from the current MicroShift container footprint?

Author: A quick `df` before and after installing MicroShift on a fresh system shows an image footprint of just over 2Gb. So this is an increase of about an 18% on-disk footprint.

logic behind this interface deploys components in an order that respects interdependencies, waits for services to 
start, and intelligently stops services on interrupt.  MicroShift's default CSI storage service is already managed 
by the service manager. 

MicroShift's CSI service manager will deploy the CSI Volume Snapshot components.  The files will be stored under the
`microshift/assets/components/csi-external-snapshot-controller` directory.  The following files will be added: 

1. `csi_controller_deployment.yaml` specifies the CSI Snapshot Controller
2. `serviceaccount.yaml` specifies the controller's service account
3. `05_operand_rbac.yaml` provides RBAC configuration to the CSI controller 
4. `webhook_config.yaml`
5. `webhook_deployment.yaml`
6. `volumesnapshotclasses_snapshot.storage.k8s.io.yaml` defines the VolumeSnapshotClass CRD 
7. `volumesnapshotcontents_snapshot.storage.k8s.io.yaml` defines the VolumeSnapshotContents CRD
8. `volumesnapshots_snapshot.storage.k8s.io.yaml` defines the VolumeSnapshot CRD

### Greenboot Changes

MicroShift's greenboot scripts will be changed to include checks for the CSI Snapshot Controller and 

> WebhookValidation Pod.

Reviewer 2: Let's add specifics here. Changing the "expected" pod namespace and count is done in the following place.
https://github.com/openshift/microshift/blob/809c1b9182aa04a9d40ac101ed60cb0d7b6a9d09/packaging/greenboot/microshift-running-check.sh#L6

cluster workloads.  The snapshot controller, along with the CSI external snapshot sidecar, will provide an API 
driven pattern for managing stateful workload data.

## Motivation

CSI snapshot functionality was originally excluded from the CSI driver integration in MicroShift to support the 
low-resource overhead goals of the project. However, user feedback has made it clear that a supportable, robust 
backup/restore solution is necessary.  While it would be possible to run a workflow out-of-band to manage workload 
data, this would be reinventing the wheel and contribute significantly to technical debt.  CSI Snapshots are already 
integrated into OpenShift with ongoing downstream support. 

### User Stories

As an edge device owner running stateful workloads on MicroShift, I want to create in-cluster snapshots of that state 
and to restore workloads to that state utilizing existing Kubernetes patterns.

### Goals

* Enable an in-cluster workflow for snapshotting and restoration of cluster workload data
* Follow the [MicroShift design principles](https://github.com/openshift/microshift/blob/main/docs/design.md)

> ### Non-Goals

Reviewer 3: Please explicitly mention that snapshots of the root disk / partition are out of scope of this enhancement. Only PVs / PVCs can be snapshotted.

I can easily imagine that an app on an edge cluster has bunch of HostPath volumes using the host's root filesystem to configure the host or store some data, this cannot be snapshotted.

_Workflow_

1. The user creates a PVC with the following fields:
   - `spec.storageClassName`: must be the same storage class that provisioned the source volume
   - `spec.dataSource.name`: name of the source `volumeSnapshot`
   - `spec.dataSource.kind`: VolumeSnapshot
   - `spec.dataSource.apiGroup`: snapshot.storage.k8s.io
2. The user creates a Pod to consume the PVC. Required because LVMS only supports `WaitForFirstConsumer` provisioning.
3. The _CSI external provisioner_ sidecar, already integrated into the _topolvm-controller_, detects the PVC request,
   and makes a `CreateVolume` gRPC call to LVMS. 
4. _topolvm-controller_ creates a `logicalVolume.topolvm.io` CRD with the same values of the source `logicalVolume.
   topolvm.io` instance. This ensures the new volume is created on the same node as the source.
5. _topolvm-node_ creates a new thin-volume from the volume snapshot.
6. _topolvm-controller_ returns the success status to the `CreateVolume` gRPC caller.
7. The _CSI external provisioner_ creates a `persistentVolume` to track the new thin-volume
8. The _CSI external provisioner_ binds the PVC and PV together
9. The workload attaches the new volume and starts

> #### Volume Cloning

Reviewer 3: Technically, you don't need Kubernetes snapshots (CRDs, external-snapshotter sidecar, snapshot-controller), all the handling below is implemented in the external-provisioner and topolvm CSI driver that you already have.

Reviewer: Does that mean we don't need to bring in some of the components mentioned in this doc? What API would a user use to trigger a snapshot?

Reviewer 3: In Kubernetes, user asks directly in a PVC to provision a new volume as a clone of another PVC in the same namespace.
https://github.com/kubernetes/enhancements/blob/26c0a72b9b216dc86a6e02bcdf96eb9f6afce0d3/keps/sig-storage/989-extend-datasource/README.md?plain=1#L71-L73

It's up to the CSI driver what it does with such request, in LVM case it creates a snapshot under the hood. But the snapshot is not exposed in Kubernetes API.

Reviewer: OK, the user we're working with wants to integrate with a specific backup tool. @copejon , let's see if we can find out if that tool can use the approach @jsafrane describes here.

Author: I'll pull the cloning section since it's out of scope.  @dhellmann I'll update the microshift docs with it instead.

guaranteed backwards compatible within each major version.  Therefore, micro-version upgrades will not require user 
intervention to protect snapshots from deletion or orphaning.

Upgrading the CSI containers will introduce a small interrupt window to provisioning operations.  Therefore, 
to mitigate any risk or orphaning resources, it is recommended that snapshotting be allowed to complete before 
restarting MicroShift and initiating the upgrade.

#### Configuring

Cluster admins will use the VolumeSnapshotClass API to set dynamic configurations for snapshot creation. 
Driver-specific configuration is available through the ‘parameters’ sub-field, which is a string:string map and is 
defined by the particular storage provider.

#### Deploying Applications

### API Extensions

CSI Volume Snapshot APIs are a core component of OpenShift Container Platform and are detailed in
[OCP documentation](https://docs.openshift.com/container-platform/4.13/storage/container_storage_interface/persistent-storage-csi-snapshots.html).

A portion of the OpenShift documentation is provided below for additional API detail.

> #### VolumeSnapshotClass

Reviewer 3: I am not sure listing the CRDs is useful, a link to doc should be enough.

- Increased overhead.  The CSI snapshot controller and sidecar increase the on-disk footprint by a
total of 360MB.  At idle, the components use a negligible amount of CPU (>1% CPU on a 2 core machine)
and roughly 18Mb.

### Drawbacks

## Design Details

### CSI Snapshotter Components

The total additional cluster components are:

1. VolumeSnapshot CRD
2. VolumeSnapshotContent CRD
3. VolumeSnapshotClass CRD

### CSI-Snapshot Controller

The CSI Snapshot Controller is maintained SIG-Storage upstream. Downstream OCP releases of this container 
image will be used by MicroShift.  The controller is a standalone component responsible for watching 
`volumeSnapshot`, `volumeSnapshotContent`, and `volumeSnapshotClass` objects.  When a `volumeSnapshot` is created, 
the controller will trigger a snapshot operation by creating a `volumeSnapshotConent` object, which will be 
detected by the _CSI external-snapshotter sidecar_.

MicroShift will embed the controller's manifests and deploy them during startup.  They are: 

1. CSI VolumeSnapshot Controller Deployment
2. CSI VolumeSnapshot Controller Service Account
3. CSI VolumeSnapshot Controller ClusterRole and ClusterRoleBinding

> 4. CSI VolumeSnapshot Controller Role and RoleBinding

Reviewer 3: this should deploy also snapshot CRDs, https://github.com/openshift/csi-external-snapshotter/tree/master/client/config/crd

Reviewer 3: I can see you mention the CRDs in MicroShift Assets below and basically repeat the same as here, only with file names. Is it useful?

Author: Thanks, good point. will fix.

### CSI-External

The CSI Snapshot Validation Webhook is maintained SIG-Storage upstream. Downstream OCP releases of this container 
image will be used by MicroShift. The container image is specified by LVMS and is included in the _topolvm-controller_ 
deployment. It watches for `volumeSnapshotContent` and `volumeSnapshotClass` events.  When an event is detected, it 
makes the appropriate CSI gRPC call to LVMS via a shared unix socket.

### CSI-Validation-Webhook

The CSI Snapshot Validation Webhook is maintained SIG-Storage upstream. Downstream OCP releases of this container 
image will be used by MicroShift.  The webhook serves as gatekeeper to `volumeSnapshot` CREATE and UPDATE events. 
For the specifics of validation, _see_ 

> [Kubernetes KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/1900-volume-snapshot-validation-webhook#validating-scenarios).

Reviewer 3: both these chapter describe CSI Snapshot Validation Webhook deployment. It's odd to deploy it as part of _topolvm-controller_ deployment, so probably only `### CSI-Validation-Webhook` should remain.

Author: That's my mistake.  I should have been more clear that it's meant to be deployed discretely.

Reviewer 3: > Also, is the webhook deployed by CSI Snapshot Controller Operator

Yes.

   - The `startCSIPlugin()` starts LVMS immediately after. It is not necessary to wait for the CSI Controller to 
         become **Ready** as there are no start-sequence dependencies between the 2 systems.
4. The CSI VolumeSnapshot Controller begins listening for VolumeSnapshot API events.
5. The user observes the CSI Snapshot Controller pod reaches the **Ready** state.
   - `$ oc get pod -n kube-system csi-snapshot-controller-$SUFFIX`
6. The user observes the `topolvm-controller` and `csi-snapshot-controller` pods reach the **Ready** state.
    - `$ oc get pod -n openshift-storage topolvm-controller-$SUFFIX`
    - The `topolvm-controller` includes the `csi-external-snapshotter` sidecar container, which is why the user must 
      verify its phase.
      
#### Create a Snapshot Dynamically from a PVC

This section describes how to create snapshot from a PVC utilizing a VolumeSnapshotClass to dynamically specify 
provisioning parameters.

_Prerequisites_

* A running MicroShift cluster
* CSI Snapshot Controller is in **Ready** state
* Topolvm-Controller is in **Ready** state
* `volumeSnapshotClass` has been deployed
* `storageClass` has been deployed

_Workflow_
 

> 1. The user creates a PVC of an arbitrary size, in Gb increments, to be the source volume from which we will 

Author: Need to make it clear that 1-5 are preliminary steps a user would follow to deploy a workload.

   create a snapshot.
2. The user creates a Pod which consumes the PVC.  This is required because LVMS only supports 
   `WaitForFirstConsumer` provisioning.
3. LVMS provisions the backend storage volume and creates a `logicalVolume.topolvm.io` CR.
4. The _csi-external-provisioner_, part of the _topolvm-controller_ pod, binds the PVC to the volume's PV.

> 5. The user's workload starts and mounts the volume.

Reviewer 2: Do we want to make it more explicit that steps 1-5 are user-pod-allocation workflow related, and not part of the snapshot taking workflow.
